res/controllers/common-hid-packet-parser.js

```diff
@@ -1,1549 +1,2310 @@
+     */
+    constructor(name, reportId = 0, callback = undefined, header = []) {
+        /**
+         * Name of packet
+         *
+         * @type {string}
+         */
+        this.name = name;
+
+        /**
+         * ReportID of the packet. If the device does not use ReportIDs this must be 0.
+         *
+         * @type {number}
+         */
         this.reportId = reportId;
-    }
 
-    this.groups = {};
+        /**
+         * Function to call when the packet type represents an InputReport, and a new report is received.
+         *
+         * @type {packetCallback}
+         */
+        this.callback = callback;
 
-    // Size of various 'pack' values in bytes
-    this.packSizes = {b: 1, B: 1, h: 2, H: 2, i: 4, I: 4};
-    this.signedPackFormats = ["b", "h", "i"];
-};
+        /**
+         * List of bytes to match from beginning of packet
+         *
+         * @type {number[]}
+         */
+        this.header = header;
 
-/** Pack a field value to the packet.
- * Can only pack bits and byte values, patches welcome. */
-HIDPacket.prototype.pack = function(data, field) {
-    var value;
-    if (!(field.pack in this.packSizes)) {
-        HIDDebug("ERROR parsing packed value: invalid pack format " + field.pack);
-        return;
-    }
-    var bytes = this.packSizes[field.pack];
-    var signed = false;
-    if (this.signedPackFormats.indexOf(field.pack) !== -1)
-        signed = true;
-
-    if (field.type === "bitvector") {
-        // TODO - fix multi byte bit vector outputs
-        if (bytes > 1) {
-            HIDDebug("ERROR: packing multibyte bit vectors not yet supported");
+        this.groups = {};
+
+        /**
+         * Length of packet in bytes
+         *
+         * @type {number}
+         */
+        this.length = this.header.length;
+
+        // Size of various 'pack' values in bytes
+        this.packSizes = {b: 1, B: 1, h: 2, H: 2, i: 4, I: 4};
+        this.signedPackFormats = ["b", "h", "i"];
+    }
+
+    /**
+     * Pack a field value to the packet.
+     * Can only pack bits and byte values, patches welcome.
+     *
+     * @todo Implement multi byte bit vector outputs
+     * @param {Uint8Array} data Data received as InputReport from the device
+     * @param {packetField} field Object that describes a field inside of a packet, which can often
+     *     mapped to a Mixxx control.
+     */
+    pack(data, field) {
+        if (!(field.pack in this.packSizes)) {
+            console.error("HIDPacket.pack - Parsing packed value: invalid pack format " + field.pack);
             return;
         }
-        for (var bit_id in field.value.bits) {
-            var bit = field.value.bits[bit_id];
-            data[field.offset] = data[field.offset] | bit.value;
+        const bytes = this.packSizes[field.pack];
+        const signed = this.signedPackFormats.includes(field.pack);
+        if (field.type === "bitvector") {
+            const bitVector = /** @type {HIDBitVector} */ (field.value);
```

Comment:
This is not the same. The `/** @type {HIDBitVector} */` right of the `=` is a typecast, if it would a line before, it would be a type definition for the constant bitVector.
field.value can be of type number, boolean or HIDBitVector. Because it's verified in the line before, that it's a bitvector, we can safely typecast here.

Reply:
I didn't know typecasts were a thing in JSDoc. Ignore my suggestion.